### PR TITLE
feat: implement field_columns plan

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/810 (includes rationale)

# Changes
This PR ports the `field_columns` planning logic to the `InfluxRPC` frontend, along with tests, so that it now works across mutable buffer and read buffer.

# Things I am pleased with:
* it was straightforward to port the code and tests
* The actual new logic is very small and the gRPC planner has almost all the right logic already
* requires no changes to the read buffer (or mutable buffer, other than removing code)

# Things I am not pleased with:

The performance of the implementation of `field_columns` is likely from ideal. The only difference between the tag_names and field_columns is that field_columns returns type information as well as a name and a timestamp.

However, the  logic in the mutable buffer (and hence now in the gRPC planner) basically runs  `SELECT <field columns> from TABLE where PREDICATE` for each table specified. The executor scans the output of those plans and finds which columns are null. It would be cool to have special handling for this kind of predicate in the `MutableBuffer` and `ReadBuffer`

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: server/src/query_tests/scenarios.rs [hunk 1]
```diff
@@ -78,51 +78,7 @@ impl DBSetup for TwoMeasurements {
                     cpu,region=west user=21.0 150\n\
                     disk,region=east bytes=99i 200";
 
-        let db = make_db();
```
Comment (alamb) [inline server/src/query_tests/scenarios.rs:81]: I pulled the code that put chunks in different locations into their own functions - `make_one_chunk_scenarios` and `make_two_chunk_scenarios` to reduce some copy/paste (that I was about to copy/paste again)

File: server/src/query_tests/scenarios.rs [hunk 2]
```diff
@@ -212,78 +104,58 @@ pub struct TwoMeasurementsManyNulls {}
...
 
-        db.load_chunk_to_read_buffer(partition_key, 1)
-            .await
-            .unwrap();
-        db.drop_mutable_buffer_chunk(partition_key, 1)
-            .await
-            .unwrap();
+        // Can't use all one chunks scenarios until read buffer supports string fields
+        //
+        // ---- query_tests::influxrpc::field_columns::test_field_name_plan stdout ----
```
Comment (alamb) [inline server/src/query_tests/scenarios.rs:150]: @e-dard  FYI I have a ready made integration test for when the ReadBuffer starts supporting string fields :) Filed https://github.com/influxdata/influxdb_iox/issues/820 to track

File: query/src/plan/fieldlist.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+use arrow_deps::datafusion::logical_plan::LogicalPlan;
+
+/// A plan which produces a logical set of Fields (e.g. InfluxDB
+/// Fields with name, and data type, and last_timestamp).
```
Comment (e-dard) [inline query/src/plan/fieldlist.rs:4]: Do the storage APIs return the timestamps to the clients? I haven't looked closely, but at least from the end-user experience you don't see timestamps when you run `SHOW FIELD KEYS` or the flux equivalent.

**Review by e-dard — APPROVED** [review]
LGTM 👍

File: query/src/plan/fieldlist.rs (hunk 3 above)
Comment (alamb) [reply query/src/plan/fieldlist.rs:4]: Yes, `MeasurementFieldResponse` returns a timestamp with each field column
https://github.com/influxdata/influxdb_iox/blob/main/generated_types/storage_common.proto#L223-L241

If we can convince ourselves that field is not used, then we can make these plans much faster (we can route them through the `tag_columns` code and answer most via metadata)

File: server/src/query_tests/scenarios.rs (hunk 2 above)
Comment (alamb) [reply server/src/query_tests/scenarios.rs:150]: @e-dard  also already has a fix :) https://github.com/influxdata/influxdb_iox/issues/822

File: query/src/plan/fieldlist.rs (hunk 3 above)
Comment (e-dard) [reply query/src/plan/fieldlist.rs:4]: @alamb I'm already ahead of you on that front :-) So I have been adding a simple argument to `column_names` in the read buffer that will let us use it for getting "tag keys" or getting "field keys". This saves having to add special case implementations for `field keys` in the read buffer.

The stickler was noting the presence of `timestamp`, which surprised me (and yeah I confirmed it's in the storage RPCs in Influx). I'm not at all convinced this is used for anything. I'm going to ask around a bit more and then get back to you on this.

Comment (alamb) [reply query/src/plan/fieldlist.rs:4]: nice! @e-dard  note that with `Schema` information the query layer already knows what columns are fields. Maybe we could just add a `Selection` argument to `read_buffer::Database::column_names`

**Comment by alamb** [comment]
I am going to merge this in -- if anyone else has any comments or suggestions, please let me know and I'll address them as a follow on PR

**Comment by alamb** [comment]
FYI @NGA-TRAN

**Pull request merged**